File: src/main/javascript/crypto/src/libs/UNMac.ts
import {Hmac} from 'crypto';
import {bnToUint8, base64ToUint8array, isDomainValid, logger, uint8arrayToBase64, base64toBase64Url} from "./utils";
import {UnpredictableNumberBundle} from "./UnpredictableNumberBundle";
import {DEBUGLEVEL} from "../config";
import { BYTES_IN_SEED, BYTES_IN_UN, hashContext, IUnpredictableNumberTool } from './IUnpredictableNumberTool';
import { TextEncoder } from 'util';

export class UNMac implements IUnpredictableNumberTool {

    private readonly _domain: string;
    private readonly validityInMs: bigint;
    private readonly key: Uint8Array;

    constructor(key: Uint8Array, domain: string, validityInMs: bigint) {
        if (!isDomainValid(domain)) {
        throw Error('Domain is not a valid domain');
        }
        this.key = key;
        this._domain = domain;
        this.validityInMs = validityInMs;
    }

    get domain(): string {
        return this._domain;
    }

    get unpredictableNumberBundle(): UnpredictableNumberBundle {
        return this.getUnpredictableNumberBundle();
    }

    getUnpredictableNumberBundle(context?: Uint8Array): UnpredictableNumberBundle {
        const expiration: bigint = BigInt(Date.now()) + this.validityInMs;
        const randomness = require('secure-random').randomUint8Array(BYTES_IN_SEED);
        let unpredictableNumber = this.getUnpredictableNumber(randomness, expiration, context);
        return new UnpredictableNumberBundle(unpredictableNumber, randomness, this._domain, expiration);
    }

    private getUnpredictableNumber(randomness: Uint8Array, expirationInMs: bigint, context: Uint8Array|undefined, unSize: number = BYTES_IN_UN): string {
        const hmac: Hmac = require('crypto').createHmac('sha3-256', this.key);
        // We encode this as a long of 8 bytes
        let byteTime = new Uint8Array(8);
        byteTime.set(bnToUint8(expirationInMs), 8-bnToUint8(expirationInMs).length);
        hmac.update(byteTime);
        hmac.update(randomness);
        // if (context !== undefined) {
        if (typeof context !== "undefined") {
            hmac.update(hashContext(context));
        }
        let encodedDomain = new TextEncoder().encode(this._domain);
        hmac.update(encodedDomain);
        const digest: Buffer = hmac.digest();
        const result = digest.slice(0, unSize);
        return base64toBase64Url(uint8arrayToBase64(result));
   }

    validateUnpredictableNumber(un: string, randomness:Uint8Array, expirationInMs: bigint, context?:Uint8Array): boolean {
        if ( BigInt(Date.now()) > expirationInMs) {
            logger(DEBUGLEVEL.LOW, 'Unpredictable number has expired');
            return false;
        }
        const expectedNumber = this.getUnpredictableNumber(randomness, expirationInMs, context, base64ToUint8array(un).length);
        if (expectedNumber !== un) {
            logger(DEBUGLEVEL.LOW, 'The unpredictable number is computed incorrectly. Either wrong key or wrong domain');
            return false;
        }
        return true;
    }
  
}

File: src/main/java/org/tokenscript/attestation/core/UNMac.java
package org.tokenscript.attestation.core;

import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.Clock;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.crypto.digests.SHA3Digest;
import org.bouncycastle.crypto.macs.HMac;
import org.bouncycastle.crypto.params.KeyParameter;
import org.tokenscript.eip712.Eip712Common;

public class UNMac implements UnpredictableNumberTool {

  public static final int BYTES_IN_UN = 16; // 128 bits
  private static final Logger logger = LogManager.getLogger(UNMac.class);

  private final SecureRandom random;
  private final String domain;
  private final long validityInMs;
  private final HMac hmac = new HMac(new SHA3Digest(BYTES_IN_SEED * 8));

  public UNMac(byte[] key, String domain) {
    this(new SecureRandom(), key, domain);
  }

  public UNMac(SecureRandom random, byte[] key, String domain) {
    this(random, key, domain, DEFAULT_VALIDITY_IN_MS);
  }

  public UNMac(SecureRandom random, byte[] key, String domain, long validityInMs) {
    this.random = random;
    this.domain = domain;
    this.validityInMs = validityInMs;
    hmac.init(new KeyParameter(key));
    // todo should be moved to url utility
    if (!Eip712Common.isDomainValid(domain)) {
      throw ExceptionUtil.throwException(logger, new IllegalArgumentException("Domain is not a valid domain"));
    }
  }

  @Override
  public String getDomain() {
    return domain;
  }

  @Override
  public UnpredictableNumberBundle getUnpredictableNumberBundle() {
    return getUnpredictableNumberBundle(null);
  }


  @Override
  public UnpredictableNumberBundle getUnpredictableNumberBundle(byte[] context) {
    long expiration = Clock.systemUTC().millis() + validityInMs;
    byte[] randomness = random.generateSeed(BYTES_IN_SEED);
    // Construct UN of BYTES_IN_UN bytes
    return new UnpredictableNumberBundle(
        getUnpredictableNumber(randomness, expiration, context, BYTES_IN_UN)
        , randomness, domain, expiration, context);
  }

  private String getUnpredictableNumber(byte[] randomness, long expirationInMs, byte[] context,
      int unSize) {
    // compute HMAC on the expiration, randomness and the hash digest of the context
    hmac.reset();
    hmac.update(UnpredictableNumberTool.longToBytes(expirationInMs), 0, Long.BYTES);
    hmac.update(randomness, 0, BYTES_IN_SEED);
    if (context != null) {
      hmac.update(UnpredictableNumberTool.hashContext(context), 0, BYTES_IN_SEED);
    }
    hmac.update(domain.getBytes(StandardCharsets.UTF_8), 0,
        domain.getBytes(StandardCharsets.UTF_8).length);
    byte[] digest = new byte[BYTES_IN_SEED];
    hmac.doFinal(digest, 0);
    byte[] result = new byte[unSize];
    System.arraycopy(digest, 0, result, 0, unSize);
    return URLUtility.encodeData(result);
  }

  @Override
  public boolean validateUnpredictableNumber(String un, byte[] randomness, long expirationInMs) {
    return validateUnpredictableNumber(un, randomness, expirationInMs, null);
  }

  @Override
  public boolean validateUnpredictableNumber(String un, byte[] randomness, long expirationInMs, byte[] context) {
    if (Clock.systemUTC().millis() > expirationInMs) {
      logger.error("Unpredictable number has expired");
      return false;
    }
    int unByteLength = URLUtility.decodeData(un).length;
    String expectedNumber = getUnpredictableNumber(randomness, expirationInMs, context,
        unByteLength);
    if (!expectedNumber.equals(un)) {
      logger.error(
          "The unpredictable number is computed incorrectly. Either wrong key or wrong domain");
      return false;
    }
    return true;
  }
}
